pkg/webhook/utils.go

```diff
@@ -111,3 +116,117 @@ func ValidateSingletonCreation(ctx context.Context, cli client.Reader, req *admi
 	return DenyCountGtZero(ctx, cli, resourceGVK,
 		fmt.Sprintf("Only one instance of %s object is allowed", req.Kind.Kind))
 }
+
+// ValidateConnectionAnnotation validates the connection annotation  "opendatahub.io/connections"
+// If the annotation exists and has a non-empty value, it validates that the value references
+// a valid secret in the same namespace. Additionally, it checks the secret's connection type
+// annotation and rejects requests with invalid configurations. (see allowedTypes)
+// If the annotation doesn't exist or is empty, it allows the operation.
+//
+// Parameters:
+//   - ctx: Context for the API call (logger is extracted from here).
+//   - cli: The controller-runtime reader to use for getting secrets.
+//   - decoder: The admission decoder to decode the request object.
+//   - req: The admission request being processed.
+//   - allowedTypes: List of allowed connection types for validation.
+//
+// Returns:
+//   - admission.Response: The validation result
+//   - bool: true if injection should be performed (only for known valid connection types)
+//   - *corev1.Secret: The validated secret (only valid when injection should be performed)
+//   - string: The connection type (only valid when injection should be performed)
+func ValidateConnectionAnnotation(ctx context.Context,
+	cli client.Reader,
+	decoder admission.Decoder,
+	req admission.Request,
+	allowedTypes []string,
+) (admission.Response, bool, *corev1.Secret, string) {
+	log := logf.FromContext(ctx)
+
+	// Decode the InferenceService object from the request
+	obj := &unstructured.Unstructured{}
+	if err := decoder.Decode(req, obj); err != nil {
+		log.Error(err, "failed to decode InferenceService object")
+		return admission.Errored(http.StatusInternalServerError, err), false, nil, ""
+	}
+
+	// Get annotations from the request	object
+	objAnnotations := obj.GetAnnotations()
+	if objAnnotations == nil {
+		objAnnotations = make(map[string]string)
+	}
+
+	// Check if the annotation "opendatahub.io/connections" exists and has a non-empty value
+	annotationValue, exists := objAnnotations[annotations.Connection]
+	if !exists || annotationValue == "" {
+		// If annotation doesn't exist or is empty, allow the operation skip injection
+		return admission.Allowed(fmt.Sprintf("Annotation '%s' not present or empty value, skipping validation", annotations.Connection)), false, nil, ""
+	}
+
+	// If annotation exists and has a value, validate the secret value
+	secret := &corev1.Secret{}
+	secretKey := types.NamespacedName{
+		Name:      annotationValue,
+		Namespace: req.Namespace,
+	}
+	if err := cli.Get(ctx, secretKey, secret); err != nil {
```

Comment:
since you don't need the entire secret content, you can probably sue `PartialObjectMetadata` to reduce the amount of data transferred over the network,